### PR TITLE
define auto remedial action if time to implement is below threshold

### PR DESCRIPTION
**Pull request by jipea**
**Please check if the PR fulfills these requirements**

- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

**What kind of change does this PR introduce?**

**What is the current behavior?**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

**Other information**:

### CODE REVIEW

File: data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java [hunk 1]
```diff
@@ -56,4 +67,66 @@ void importPstRangeActions() {
...
+        assertFlowCnecEquality(importedFlowCnecs.get(1), "RTE_AE2 (ae-2) - RTE_CO - curative", "FFR2AA1  FFR3AA1  1",
+            CURATIVE_INSTANT_ID, "contingency", null, null, 2500.0, -2500.0, Set.of(Side.RIGHT), "RTE");
+        assertFlowCnecEquality(importedFlowCnecs.get(2), "RTE_AE2 (ae-2) - preventive", "FFR2AA1  FFR3AA1  1",
+            PREVENTIVE_INSTANT_ID, null, null, null, 2500.0, -2500.0, Set.of(Side.RIGHT), "RTE");
+
+        List<PstRangeAction> importedPstRangeActions = cracCreationContext.getCrac().getPstRangeActions().stream().sorted(Comparator.comparing(PstRangeAction::getId)).toList();
+        assertEquals(3, importedPstRangeActions.size());
+
+        assertPstRangeActionImported(cracCreationContext, "auto-pst-be2-be3", "BBE2AA1  BBE3AA1  1", false, 1, null);
+        assertHasOnFlowConstraintUsageRule(cracCreationContext, "auto-pst-be2-be3", "RTE_AE1 (ae-1) - RTE_CO - auto - TATL 900", cracCreationContext.getCrac().getInstant(AUTO_INSTANT_ID), UsageMethod.AVAILABLE);
```
Comment (bqth29) [inline data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java:95]: The UsageMethod should be FORCED
You need to refactor the addOnFlowConstraint and adOnContingencyState methods in order to force the PST RA if it is at the auto instant

File: data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java [hunk 2]
```diff
@@ -56,4 +67,66 @@ void importPstRangeActions() {
...
+        assertFlowCnecEquality(importedFlowCnecs.get(1), "RTE_AE2 (ae-2) - RTE_CO - curative", "FFR2AA1  FFR3AA1  1",
+            CURATIVE_INSTANT_ID, "contingency", null, null, 2500.0, -2500.0, Set.of(Side.RIGHT), "RTE");
+        assertFlowCnecEquality(importedFlowCnecs.get(2), "RTE_AE2 (ae-2) - preventive", "FFR2AA1  FFR3AA1  1",
+            PREVENTIVE_INSTANT_ID, null, null, null, 2500.0, -2500.0, Set.of(Side.RIGHT), "RTE");
+
+        List<PstRangeAction> importedPstRangeActions = cracCreationContext.getCrac().getPstRangeActions().stream().sorted(Comparator.comparing(PstRangeAction::getId)).toList();
+        assertEquals(3, importedPstRangeActions.size());
+
+        assertPstRangeActionImported(cracCreationContext, "auto-pst-be2-be3", "BBE2AA1  BBE3AA1  1", false, 1, null);
+        assertHasOnContingencyStateUsageRule(cracCreationContext, "auto-pst-be2-be3", "contingency", AUTO_INSTANT_ID, UsageMethod.FORCED);
```
Comment (bqth29) [inline data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java:126]: Same comment about FORCED

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/parameters/CsaCracCreationParameters.java [hunk 3]
```diff
@@ -27,4 +28,12 @@ public String getCapacityCalculationRegionEicCode() {
...
         this.capacityCalculationRegionEicCode = capacityCalculationRegionEicCode;
     }
+
+    public Integer getSpsMaxTimeToImplementThresholdInSeconds() {
+        return spsMaxTimeToImplementThresholdInSeconds;
+    }
+
+    public void setSpsMaxTimeToImplementThresholdInSeconds(Integer spsMaxTimeToImplementThresholdInSeconds) {
+        this.spsMaxTimeToImplementThresholdInSeconds = spsMaxTimeToImplementThresholdInSeconds;
+    }
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/parameters/CsaCracCreationParameters.java]: done

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/parameters/JsonCsaCracCreationParameters.java [hunk 4]
```diff
@@ -23,11 +23,13 @@
 public class JsonCsaCracCreationParameters implements JsonCracCreationParameters.ExtensionSerializer<CsaCracCreationParameters> {
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/parameters/JsonCsaCracCreationParameters.java:24]: I've updated existing test cases in JsonCsaCracCreationParametersTest.java

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/CsaProfileCracCreator.java [hunk 5]
```diff
@@ -50,6 +50,7 @@ public String getNativeCracFormat() {
     public CsaProfileCracCreationContext createCrac(CsaProfileCrac nativeCrac, Network network, OffsetDateTime offsetDateTime, CracCreationParameters cracCreationParameters) {
         CsaCracCreationParameters csaParameters = cracCreationParameters.getExtension(CsaCracCreationParameters.class);
         String regionEicCode = csaParameters.getCapacityCalculationRegionEicCode();
+        Integer spsMaxTimeToImplementThreshold = csaParameters.getSpsMaxTimeToImplementThresholdInSeconds();
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/CsaProfileCracCreator.java:53]: done

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java [hunk 6]
```diff
@@ -43,23 +43,23 @@ public class CsaProfileRemedialActionsCreator {
...
+        createRemedialActions(true, spsMaxTimeToImplementThreshold);
         // standaloneRaIdsImplicatedIntoAGroup contain ids of Ra's depending on a group whether the group is imported or not
         Set<String> standaloneRaIdsImplicatedIntoAGroup = createRemedialActionGroups();
         standaloneRaIdsImplicatedIntoAGroup.forEach(crac::removeRemedialAction);
         standaloneRaIdsImplicatedIntoAGroup.forEach(importedRaId -> contextByRaId.remove(importedRaId));
         this.cracCreationContext.setRemedialActionCreationContexts(new HashSet<>(contextByRaId.values()));
     }
 
-    private void createRemedialActions(boolean isAuto) {
+    private void createRemedialActions(boolean isAuto, int spsMaxTimeToImplementThreshold) {
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java]: done

File: data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/RemedialActionCreationTest.java [hunk 7]
```diff
@@ -26,7 +29,8 @@ class RemedialActionCreationTest {
 
     @Test
     void importRemedialActions() {
-        CsaProfileCracCreationContext cracCreationContext = getCsaCracCreationContext("/profiles/remedialactions/RemedialActions.zip", NETWORK);
+        CracCreationParameters importedParameters = JsonCracCreationParameters.read(getClass().getResourceAsStream("/parameters/csaCracCreationParameters_SPS_60_sec.json"));
+        CsaProfileCracCreationContext cracCreationContext = getCsaCracCreationContext("/profiles/remedialactions/RemedialActions.zip", NETWORK, importedParameters);
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/RemedialActionCreationTest.java]: ok

File: data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java (hunk 1 above)
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java]: ok

File: data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java (hunk 2 above)
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/PstRangeActionCreationTest.java:126]: ok

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java [hunk 8]
```diff
@@ -112,8 +112,22 @@ private RemedialActionAdder<?> getRemedialActionAdder(String remedialActionId, S
...
+                try {
+                    durationInSeconds = CsaProfileCracUtils.convertDurationToSeconds(timeToImplement);
+                } catch (RuntimeException e) {
+                    throw new OpenRaoImportException(ImportStatus.INCONSISTENCY_IN_DATA, "Remedial action " + remedialActionId + " will not be imported because of an irregular timeToImplement pattern");
+                }
+                if (durationInSeconds <= durationLimit) {
+                    instant = crac.getInstant(InstantKind.AUTO);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java]: done

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java [hunk 9]
```diff
@@ -258,6 +272,27 @@ private void addOnContingencyStateUsageRules(PropertyBag parentRemedialActionPro
         }
     }
 
+    private String defineInstantId(boolean isAuto, PropertyBag parentRemedialActionPropertyBag, String remedialActionId, int durationLimit) {
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java]: done

File: data/crac-creation/crac-creator-csa-profiles/src/test/resources/parameters/csa-crac-parameters-json-csa-crac-creation-parameters-test.json [hunk 10]
```diff
@@ -0,0 +1,9 @@
+{
+  "crac-factory" : "CracImplFactory",
+  "extensions" : {
+    "CsaCracCreatorParameters" : {
+      "capacity-calculation-region-eic-code": "10Y1001C--00095L-1",
```
Comment (bqth29) [inline data/crac-creation/crac-creator-csa-profiles/src/test/resources/parameters/csa-crac-parameters-json-csa-crac-creation-parameters-test.json:5]: Can you use 10Y1001C--00095L EIC instead please? (better to use a real EI code)

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java [hunk 11]
```diff
@@ -247,27 +250,45 @@ private void addOnContingencyStateUsageRules(PropertyBag parentRemedialActionPro
...
+        }
+        if (durationInSeconds <= durationLimit) {
+            return crac.getInstant(InstantKind.AUTO);
+        } else {
+            return crac.getInstant(InstantKind.CURATIVE);
+        }
+    }
+
+    private UsageMethod getUsageMethod(ElementCombinationConstraintKind elementCombinationConstraintKind, boolean isSchemeRemedialAction, Instant instant) {
+        return isSchemeRemedialAction || ElementCombinationConstraintKind.INCLUDED.equals(elementCombinationConstraintKind) || instant.isAuto() ? UsageMethod.FORCED : UsageMethod.AVAILABLE;
```
Comment (bqth29) [inline data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java:286]: This should be the case for auto PST range actions only:
- at the auto instant, PST SPS must be FORCED
- at the auto instant, topo SPS can be either FORCED or AVAILABLE

File: data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/parameters/JsonCsaCracCreationParametersTest.java [hunk 12]
```diff
@@ -34,11 +34,12 @@ void basicData() {
...
 
     @Test
-    void deserializeParametersWithRegionEicCode() {
-        CracCreationParameters importedParameters = JsonCracCreationParameters.read(getClass().getResourceAsStream("/parameters/csa-crac-parameters-with-region-eic-code.json"));
+    void deserializeValidParameters() {
+        CracCreationParameters importedParameters = JsonCracCreationParameters.read(getClass().getResourceAsStream("/parameters/csa-crac-parameters-json-csa-crac-creation-parameters-test.json"));
         CsaCracCreationParameters csaCracCreationParameters = importedParameters.getExtension(CsaCracCreationParameters.class);
         assertNotNull(csaCracCreationParameters);
-        assertEquals("10Y1001C--00095L", csaCracCreationParameters.getCapacityCalculationRegionEicCode());
+        assertEquals("10Y1001C--00095L-1", csaCracCreationParameters.getCapacityCalculationRegionEicCode());
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/test/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/parameters/JsonCsaCracCreationParametersTest.java:41]: I change it to RTE EI code to have a real EI code but different from default EI code

File: data/crac-creation/crac-creator-csa-profiles/src/test/resources/parameters/csa-crac-parameters-json-csa-crac-creation-parameters-test.json (hunk 10 above)
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/test/resources/parameters/csa-crac-parameters-json-csa-crac-creation-parameters-test.json:5]: I change it to RTE EI code to have a real EI code but different from default EI code

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java [hunk 13]
```diff
@@ -168,12 +171,12 @@ private Optional<Integer> getSpeedOpt(RemedialActionType remedialActionType, Str
...
         } else {
             alterations.add(String.format("The association 'RemedialAction'/'Cnecs' '%s'/'%s' will be ignored because 'excluded' combination constraint kind is not supported", importableRemedialActionId, String.join(". ", onConstraintUsageRuleHelper.getExcludedCnecsByRemedialAction().get(importableRemedialActionId))));
         }
         if (onConstraintUsageRuleHelper.getConsideredAndIncludedCnecsByRemedialAction().containsKey(importableRemedialActionId)) {
-            onConstraintUsageRuleHelper.getConsideredAndIncludedCnecsByRemedialAction().get(importableRemedialActionId).forEach(addOnConstraintUsageRuleForCnec(remedialActionInstant, remedialActionAdder, UsageMethod.AVAILABLE));
+            onConstraintUsageRuleHelper.getConsideredAndIncludedCnecsByRemedialAction().get(importableRemedialActionId).forEach(addOnConstraintUsageRuleForCnec(remedialActionInstant, remedialActionAdder, remedialActionInstant.isAuto() ? UsageMethod.FORCED : UsageMethod.AVAILABLE));
```
[the remark this answers is not in the document]
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java]: ok, I've updated it to differentiate PST SPS and topo

File: data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java (hunk 11 above)
Comment (jipea) [reply data/crac-creation/crac-creator-csa-profiles/src/main/java/com/powsybl/openrao/data/craccreation/creator/csaprofile/craccreator/remedialaction/CsaProfileRemedialActionsCreator.java]: ok, I've updated it to differentiate PST SPS and topo

**Review by bqth29 — APPROVED** [review]

**Pull request merged**